Rearrange fields of ecma_property_t to be naturally aligned.

zherczeg opened the pull request:
Rearrange fields of ecma_property_t to be naturally aligned. Packed attribute and __extension__ keywords are removed. The standard approach reduced the binary size by 2K.

zherczeg commented:
Benchmark |                 Perf(+ is better)
                   3d-cube.js |   2.620s ->   2.660s :  -1.527%  (+-1.198%) : [-]
       access-binary-trees.js |   1.405s ->   1.370s :  +2.491%  (+-0.626%) : [+]
           access-fannkuch.js |   8.003s ->   7.952s :  +0.645%  (+-0.906%) : [≈]
              access-nbody.js |   3.103s ->   3.062s :  +1.342%  (+-0.824%) : [+]
  bitops-3bit-bits-in-byte.js |   1.735s ->   1.745s :  -0.576%  (+-0.737%) : [≈]
       bitops-bits-in-byte.js |   2.663s ->   2.663s :  +0.000%  (+-0.814%) : [≈]
        bitops-bitwise-and.js |   3.647s ->   3.575s :  +1.965%  (+-0.441%) : [+]
     controlflow-recursive.js |   0.890s ->   0.892s :  -0.187%  (+-0.755%) : [≈]
                crypto-aes.js |   4.623s ->   4.612s :  +0.252%  (+-0.324%) : [≈]
                crypto-md5.js |  21.475s ->  20.520s :  +4.447%  (+-0.574%) : [+]
               crypto-sha1.js |   9.740s ->   9.293s :  +4.586%  (+-0.466%) : [+]
         date-format-tofte.js |   2.222s ->   2.152s :  +3.151%  (+-0.768%) : [+]
         date-format-xparb.js |   1.057s ->   1.052s :  +0.473%  (+-1.417%) : [≈]
               math-cordic.js |   3.017s ->   3.005s :  +0.387%  (+-0.535%) : [≈]
         math-partial-sums.js |   1.662s ->   1.657s :  +0.301%  (+-0.651%) : [≈]
        math-spectral-norm.js |   1.483s ->   1.507s :  -1.573%  (+-0.817%) : [-]
              string-fasta.js |   4.555s ->   4.485s :  +1.537%  (+-0.802%) : [+]
              Geometric mean: |                 Speed up: 1.058% (+-0.192%) : [+]

akiss77 commented on jerry-core/ecma/base/ecma-helpers.c:
```diff
@@ -522,16 +519,22 @@ ecma_create_named_data_property (ecma_object_t *obj_p, /**< object */
+  }
+  if (is_writable)
+  {
+    prop_p->flags = (uint8_t) (prop_p->flags | ECMA_PROPERTY_FLAG_WRITABLE);
```
Wouldn't `prop_p->flags |= (uint8_t) ECMA_PROPERTY_FLAG_xxx;` work here and above? (It's a bit more compact.)

Actually, using a local variable to compose the flags and then copy that into `prop_p->flags` would be even nicer.

zherczeg replied:
No :( c99 int conversion issue. I cannot say how much I hate it but only ++ and -- is accepted for types smaller than it. Everything else will be a warning. I even introduced macros for them in the parser.

akiss77 replied:
Ahh, too bad. I was not aware of that limitation. OK then.

akiss77 commented on jerry-core/ecma/base/ecma-helpers.c:
```diff
@@ -568,18 +571,20 @@ ecma_create_named_accessor_property (ecma_object_t *obj_p, /**< object */
+  }
+  if (is_enumerable)
+  {
+    prop_p->flags = (uint8_t) (prop_p->flags | ECMA_PROPERTY_FLAG_ENUMERABLE);
```
Same here.

akiss77 commented on jerry-core/ecma/base/ecma-helpers.c:
```diff
@@ -861,28 +866,19 @@ void
+    ecma_free_named_accessor_property (object_p, prop_p);
+  }
+  else
+  {
```
I'd add an assert here.
`JERRY_ASSERT (prop_p->flags & ECMA_PROPERTY_FLAG_INTERNAL);`

zherczeg replied:
I was thinking about it, but the free_internal property checks it. From the backtrace it is visible where it is called.

akiss77 replied:
True. But still, IMHO it is good coding practice which helps maintainibility to state that "normally, I'd have written `else if` here but I'm sure that its predicate will always hold therefore I just go with an `else`, however, I still assert on that predicate". And what's more, it does not cost anything in release builds.

akiss77 commented on jerry-core/ecma/base/ecma-helpers.c:
```diff
@@ -1096,9 +1094,16 @@ ecma_set_property_writable_attr (ecma_property_t *prop_p, /**< property */
+  }
+  else
+  {
+    prop_p->flags = (uint8_t) (prop_p->flags & ~ECMA_PROPERTY_FLAG_WRITABLE);
```
Again, `|=` and `&=` could be used.

akiss77 commented on jerry-core/ecma/base/ecma-helpers.c:
```diff
@@ -1130,17 +1128,15 @@ ecma_set_property_enumerable_attr (ecma_property_t *prop_p, /**< property */
-
-    prop_p->u.named_accessor_property.enumerable = (is_enumerable ?
-                                                    ECMA_PROPERTY_ENUMERABLE : ECMA_PROPERTY_NOT_ENUMERABLE);
+    prop_p->flags = (uint8_t) (prop_p->flags & ~ECMA_PROPERTY_FLAG_ENUMERABLE);
```
ditto

akiss77 commented on jerry-core/ecma/base/ecma-helpers.c:
```diff
@@ -1173,17 +1162,15 @@ ecma_set_property_configurable_attr (ecma_property_t *prop_p, /**< property */
-
-    prop_p->u.named_accessor_property.configurable = (is_configurable ?
-                                                      ECMA_PROPERTY_CONFIGURABLE : ECMA_PROPERTY_NOT_CONFIGURABLE);
+    prop_p->flags = (uint8_t) (prop_p->flags & ~ECMA_PROPERTY_FLAG_CONFIGURABLE);
```
ditto

akiss77 commented on jerry-core/ecma/base/ecma-helpers.c:
```diff
@@ -1215,16 +1194,15 @@ void
   else
   {
-    prop_p->u.named_accessor_property.is_lcached = (is_lcached != 0);
+    prop_p->flags = (uint8_t) (prop_p->flags & ~ECMA_PROPERTY_FLAG_LCACHED);
```
ditto

akiss77 commented on jerry-core/ecma/builtin-objects/ecma-builtin-string-prototype.c:
```diff
@@ -1815,7 +1815,8 @@ ecma_builtin_string_prototype_object_split (ecma_value_t this_arg, /**< this arg
               ecma_property_t *index_prop_p = ecma_get_named_property (match_array_obj_p, magic_index_str_p);
 
-              ecma_number_t *index_num_p = ecma_get_number_from_value (index_prop_p->u.named_data_property.value);
+              ecma_value_t index_value = ecma_get_named_data_property_value (index_prop_p);
```
Is the `index_value` local var needed? Couldn't we just inline `ecma_get_named_data_property_value (index_prop_p)` into the line below?

zherczeg replied:
Line length (>120 char) issue.

akiss77 replied:
The two expressions can still be merged into one, only the lines have to be broken at the right places in a nicely-formatted way.

zherczeg replied:
You suggested using local variables above, and here you don't like it :) Anyway, I would prefer that myself, but I couldn't do it. Please suggest me a form which is valid according to Jerry formatting guidelines.

akiss77 replied:
I'm sure you realize the differences between the two cases. With the flags above, you are accessing a field via a pointer multiple times. There, it makes sense to use a local variable to compose the needed value and then do the indirection only once. Here, however, a local variable is introduced just to get a line within length limits...

Something like below may work with stlye checkers (however, this no-indent style looks ugly to me :/ ). There is something similar in rcs/rcs-chunked-list.c already.
```
              ecma_value_t index_value =
              ecma_get_named_data_property_value (
              ecma_get_number_from_value (
              index_prop_p->u.named_data_property.value));
```

Eventually, I'm (un)happy either way.

zherczeg replied:
This is too hard to read.

akiss77 commented on jerry-core/ecma/base/ecma-helpers.c:
```diff
@@ -472,12 +471,10 @@ ecma_find_internal_property (ecma_object_t *object_p, /**< object descriptor */
        property_p = ECMA_GET_POINTER (ecma_property_t, property_p->next_property_p))
   {
-    if (property_p->type == ECMA_PROPERTY_INTERNAL)
+    if (property_p->flags & ECMA_PROPERTY_FLAG_INTERNAL
```
I'd suggest using some more parentheses to make operator precedence explicit. (Just like in other parts of the patch.)

akiss77 commented on jerry-core/ecma/base/ecma-globals.h:
```diff
@@ -214,84 +204,78 @@ typedef enum
-      /** Attribute 'Configurable' (ecma_property_configurable_value_t) */
-      unsigned int configurable : 1;
+      /** Lower 16 bits of value */
+      uint16_t value_low;
```
With this layout, we are expecting that the value of a named data property fits into 8+16 bits. Right now, this is true, but it would be good to check somehow that `ECMA_VALUE_SIZE` (which gets calculated from at least 2 components from different header files) is less than 24. (A static assert could work, but according to its documentation, they are not to be placed in headers.)

LaszloLango commented on jerry-core/ecma/base/ecma-globals.h:
```diff
@@ -214,84 +204,78 @@ typedef enum
       uint32_t value;
     } internal_property;
-  } u;
+  } v;
```
I'd prefer a real name for this. `value`?

LaszloLango commented on jerry-core/ecma/base/ecma-globals.h:
```diff
@@ -214,84 +204,78 @@ typedef enum
+    uint8_t named_data_property_value_high;
+    /** Internal property type */
+    uint8_t internal_property_type;
+  } h;
```
ditto, `header` ?

zherczeg replied:
The problem is the names will be too long, and in some cases it is easy to reach the 120 char.

LaszloLango replied:
Ok, then leave as is.

LaszloLango commented on jerry-core/ecma/base/ecma-helpers.c:
```diff
@@ -1107,19 +1112,12 @@ ecma_set_property_writable_attr (ecma_property_t *prop_p, /**< property */
 
-    return (prop_p->u.named_accessor_property.enumerable == ECMA_PROPERTY_ENUMERABLE);
-  }
+  return (prop_p->flags & ECMA_PROPERTY_FLAG_ENUMERABLE) != 0;
```
` != 0;` looks unnecessary for me

LaszloLango replied:
Sorry, my bad. It is good.

zherczeg commented:
On Sunspider benchmark suite the currently used MaxRSS measuring script reported a large memory growth, and I decided to analyse the reason. First I created a new valgrind tool called heimdall which reports all page ranges that are accessed by an application. After that I picked bitops-bitwise-and.js which memory consumption is growth from 24K to 32K.

Here are the results:

Original:

    Page size: 4096 bytes
      Modified pages: 0x3e000 - 0x3f000 [size: 4 Kb]
      Modified pages: 0x13e000 - 0x13f000 [size: 4 Kb]
      Modified pages: 0x17e000 - 0x180000 [size: 8 Kb]
      Modified pages: 0x27f000 - 0x280000 [size: 4 Kb]
      Modified pages: 0x7dbcb000 - 0x7dbcd000 [size: 8 Kb]
    Total modified pages: 28 Kb

It turned out that the MaxRSS measuring script underestimated the real memory consumption. This is a known weakness of the script, since its precision depends on its sampling frequency.

After the patch:

    Page size: 4096 bytes
      Modified pages: 0x3d000 - 0x3e000 [size: 4 Kb]
      Modified pages: 0x13d000 - 0x13f000 [size: 8 Kb]
      Modified pages: 0x17d000 - 0x17f000 [size: 8 Kb]
      Modified pages: 0x27e000 - 0x27f000 [size: 4 Kb]
      Modified pages: 0x7d96b000 - 0x7d96d000 [size: 8 Kb]
    Total modified pages: 32 Kb

Because of the binary size reduction the compiler decreased the base address of the data section which overlaps two pages now.

However, the interesting part comes here. We change the page size to 1K (down from 4K) and measured these tests again:

Original:

    Page size: 1024 bytes
      Modified pages: 0x3e000 - 0x3e400 [size: 1 Kb]
      Modified pages: 0x13e000 - 0x13e800 [size: 2 Kb]
      Modified pages: 0x17e000 - 0x17f400 [size: 5 Kb]
      Modified pages: 0x27f000 - 0x27f400 [size: 1 Kb]
      Modified pages: 0x7dbb3c00 - 0x7dbb4800 [size: 3 Kb]
    Total modified pages: 12 Kb

After the patch:

    Page size: 1024 bytes
      Modified pages: 0x3dc00 - 0x3e000 [size: 1 Kb]
      Modified pages: 0x13dc00 - 0x13e400 [size: 2 Kb]
      Modified pages: 0x17dc00 - 0x17f000 [size: 5 Kb]
      Modified pages: 0x27ec00 - 0x27f000 [size: 1 Kb]
      Modified pages: 0x7dbddc00 - 0x7dbde800 [size: 3 Kb]
    Total modified pages: 12 Kb

We can notice that the memory consumption decreased to 12K, and that is not changed after the patch. The ranges are the same, although the base address with the patch is a bit lower than the original. So the page alignment increase the 12K memory consumption to 28-32K on Linux. A huge waste! Fortunately this does not affect the microcontrollers, our real targets, so I propose to land this patch.

LaszloLango commented:
LGTM

zherczeg commented:
Landed in b8b587f76c41fcd3b46550642a64d0fad34f5610.
